packages/aztec.js/src/proof/publicRange/index.js

```diff
@@ -0,0 +1,180 @@
+    const numNotes = 2;
+
+    // Used to check the number of input notes. Boolean argument specifies whether the
+    // check should throw if not satisfied, or if we seek to collect all errors
+    // and only throw at the end. Here, set to true - immediately throw if error
+    proofUtils.checkNumNotes(notes, numNotes, true);
+
+    proofUtils.parseInputs(notes, sender);
+    // convert z_a and z_b into BN instances if they aren't already
+    let kPublicBN;
+
+    const rollingHash = new Keccak();
+
+    if (BN.isBN(kPublic)) {
+        kPublicBN = kPublic;
+    } else {
+        kPublicBN = new BN(kPublic);
+    }
+    // Check that proof data lies on the bn128 curve
+    notes.forEach((note) => {
+        const gammaOnCurve = bn128.curve.validate(note.gamma); // checking gamma point
+        const sigmaOnCurve = bn128.curve.validate(note.sigma); // checking sigma point
+
+        if (gammaOnCurve === false || sigmaOnCurve === false) {
+            throw customError(errorTypes.NOT_ON_CURVE, {
+                message: 'A group element is not on the bn128 curve',
+                gammaOnCurve,
+                sigmaOnCurve,
+                note,
+            });
+        }
+    });
+
+    notes.forEach((note) => {
+        rollingHash.append(note.gamma);
+        rollingHash.append(note.sigma);
+    });
+
+    const blindingFactors = publicRange.constructBlindingFactors(notes, kPublicBN);
+
+    const challenge = proofUtils.computeChallenge(sender, kPublicBN, notes, blindingFactors);
+    const proofData = blindingFactors.map((blindingFactor, i) => {
+        const kBar = notes[i].k
+            .redMul(challenge)
+            .redAdd(blindingFactor.bk)
+            .fromRed();
+        const aBar = notes[i].a
+            .redMul(challenge)
+            .redAdd(blindingFactor.ba)
+            .fromRed();
+
+        return [
+            `0x${padLeft(kBar.toString(16), 64)}`,
+            `0x${padLeft(aBar.toString(16), 64)}`,
+            `0x${padLeft(notes[i].gamma.x.fromRed().toString(16), 64)}`,
+            `0x${padLeft(notes[i].gamma.y.fromRed().toString(16), 64)}`,
+            `0x${padLeft(notes[i].sigma.x.fromRed().toString(16), 64)}`,
+            `0x${padLeft(notes[i].sigma.y.fromRed().toString(16), 64)}`,
+        ];
+    });
+
+    return {
+        proofData,
+        challenge: `0x${padLeft(challenge.toString(16), 64)}`,
+    };
+};
+
+/**
+ * Encode a dividend computation transaction
+ *
+ * @method encodePublicRangeTransaction
+ * @memberof module:publicRange
+ * @param {Note[]} inputNotes input AZTEC notes
+ * @param {Note[]} outputNotes output AZTEC notes
+ * @param {Number} kPublic - public integer being compared against
+ * @param {string} senderAddress the Ethereum address sending the AZTEC transaction (not necessarily the note signer)
+ * @returns {Object} AZTEC proof data and expected output
+ */
+publicRange.encodePublicRangeTransaction = ({ inputNotes, outputNotes, kPublic, senderAddress }) => {
```

Comment:
Why are we constraining ourselves to this really un-descriptive API? In this context, I wouldn't expect anyone to have any clue what inputNotes, outputNotes, or kPublic are supposed to be. 

Private range, for example, is much more descriptive. If we go in that direction, we can also feed in a boolean to denote whether this is a < or a > range comparison.

Reply:
Agreed, changed the interface to:

```
publicRange.encodePublicRangeTransaction = async ({
    originalNote,
    publicComparison,
    senderAddress,
    isGreaterOrEqual,
    utilityNote,
}) => { }
```

The utilityNote is an optional parameter - the default is that it's generated internally and is not required as an input - but some flows require it to be possible to explicitly input